### ISSUE TITLE
Generic regex rules

### ISSUE THREAD

**Issue by keith** [realm/SwiftLint#123]
As we've been expanding our usage of SwiftLint we've added a lot of simple regex rules to conform to our style guide. I'm not sure what the long term plans are to handle custom rules but it might be nice to provide a way for users to add regex rules in their `.swiftlint.yml`. For example:

```yml
regex:
  - (^[^\\s]+\\s+@objc|@objc[^\\n_])
```

This would create multiple rules (once https://github.com/realm/SwiftLint/pull/106 is merged) at runtime.

There are a few issues I can see with this, specifically around them not being customizable. I'm not sure how the warning/error message would be generated and things like `NSRegularExpression` options wouldn't be unusable.

I just wanted to get some feedback on this idea before putting time into it. Let me know your thoughts!

**Comment by jpsim** [comment]
SwiftLint's strength IMO comes from its contextual knowledge of the AST, which is why regex-based rules also typically check the syntax type of its matched tokens. So ideally, the ability to define custom rules in the configuration file should also expose that.

How about providing more information for new rules?

```yaml
custom_rules:
  ObjCRule:
    message: @objc should be avoided
    regex: (^[^\\s]+\\s+@objc|@objc[^\\n_])
    match_tokens: identifier
    severity: warning
```

**Comment by jpsim** [comment]
Most of those fields could be optional too.

**Comment by keith** [comment]
> SwiftLint's strength IMO comes from its contextual knowledge of the AST

Totally agree. Although I do think for our use case a lot of regex doesn't really matter what section it's in like the one above, and the overhead from interacting with the syntax map isn't always worth it. This obviously doesn't apply to all linting, otherwise we'd all be using Mayday!

**Comment by jpsim** [comment]
Most of the time, you won't care about the AST other than knowing you're not in a comment or string, so it could be worth exposing a helper for that.

**Comment by scottrhoyt** [comment]
I'm about 75% of the way done with this using the framework from #391. By the time that is merged, I should shortly have a new PR to add this.

**Comment by scottrhoyt** [comment]
I've got this completed now. Will submit PR after #391. I'm a big fan because it really opens up some cool customization... including pranking your coworkers  :)
